import unittest
import datetime

from django.test import Client
from django.test import TestCase
from unittest import mock
from urllib.parse import urlparse, parse_qs
from data import api_caller
from data import legal_test_data

client = Client()


class TestLegalSearch(unittest.TestCase):

    # Test1 : OK
    @mock.patch.object(api_caller, 'load_legal_search_results')
    def test_no_query(self, load_legal_search_results):
        response = client.get('/data/legal/search/')
        assert response.status_code == 200 
        load_legal_search_results.assert_not_called()

    # Test2 : A new issue is opened https://github.com/18F/fec-cms/issues/1477
    # to address this test. commenting this test for now.
    # getting reponse.status_code=301, but expecting reponse.status_code=302
    # def test_search_type_regulations_redirects(self):
    #     response = client.get('/data/legal/search/regulations',
    #                           data={
    #                               'search': 'in kind donation',
    #                               'search_type': 'regulations'})
    #     print("::: Test2 ::: search_type_regulations_redirects :::RESPONSE :::", response.status_code)
    #     assert response.status_code == 302

    #     url = urlparse(response)
    #     query = parse_qs(url.query)
    #     assert url == 'data/legal/search/regulations/'
    #     assert 'search' in query
    #     assert 'search_type' in query
    #     assert query['search'] == ['in kind donation']
    #     assert query['search_type'] == ['regulations']

    # Test3 : OK
    @mock.patch.object(api_caller, 'load_legal_search_results')
    def test_search_universal(self, load_legal_search_results):
        load_legal_search_results.return_value = legal_test_data.legal_universal_search_results()
        response = client.get('/data/legal/search/',
                              data={
                                  'search': 'in kind donation',
                                  'search_type': 'all'})
        assert response.status_code == 200
        load_legal_search_results.assert_called_once_with('in kind donation', 'all', limit=3)

    # Test4 : This test is checking against the static data on legal_test_data.py.
    # offset value is 3 and not 0. revisit the test
    # AssertionError: Expected call: load_legal_search_results('in kind donation', 'regulations', offset=0)
    # Actual call: load_legal_search_results('in kind donation', 'regulations', limit=3)  
    @mock.patch.object(api_caller, 'load_legal_search_results')
    def test_search_regulations(self, load_legal_search_results):
        load_legal_search_results.return_value = legal_test_data.regulations_search_results()
        response = client.get('/data/legal/search/regulations/',
                              data={
                                  'search': 'in kind donation',
                                  'search_type': 'regulations'})
        assert response.status_code == 200
        load_legal_search_results.assert_called_once_with('in kind donation',
                                                          'regulations', offset=0)

    # Test5 : OK. This test works only if offset value is set
    # as a string and not integer in the assert_called_once_with. 
    @mock.patch.object(api_caller, 'load_legal_search_results')
    def test_search_pagination(self, load_legal_search_results):
        load_legal_search_results.return_value = legal_test_data.regulations_search_results()
        
        response = client.get('/data/legal/search/regulations/',
                              data={
                                  'search': 'in kind donation',
                                  'search_type': 'regulations',
                                  'offset': 20})
        assert response.status_code == 200
        load_legal_search_results.assert_called_once_with('in kind donation',
                                                          'regulations', offset='20')

    # Test 6 : OK
    @mock.patch.object(api_caller, 'load_legal_search_results')
    def test_search_statutes(self, load_legal_search_results):
        load_legal_search_results.return_value = legal_test_data.statutes_search_results()
        response = client.get('/data/legal/search/statutes/',
                              data={
                                  'search': 'in kind donation',
                                  'search_type': 'statutes'})
        assert response.status_code == 200
        load_legal_search_results.assert_called_once_with('in kind donation',
                                                          'statutes', offset=0)

    # # Test 7: OK
    @mock.patch.object(api_caller, '_call_api')
    def test_result_counts(self, _call_api_mock):
        _call_api_mock.return_value = {
            'advisory_opinions': [{'no': 1, 'date': '2016'}, {'no': 2, 'date': '1999'}],
            'statutes': [{}] * 4,
            'regulations': [{}] * 5}
        results = api_caller.load_legal_search_results(query='president')

        assert len(results['advisory_opinions']) == 2
        assert results['advisory_opinions_returned'] == 2
        assert results['statutes_returned'] == 4
        assert results['regulations_returned'] == 5

    # # Test 8: OK
    @mock.patch.object(api_caller, 'load_legal_search_results')
    def test_ao_landing_page(self, load_legal_search_results):
        today = datetime.date.today()
        ao_min_date = today - datetime.timedelta(weeks=26)
        response = client.get('/data/legal/advisory-opinions/')

        assert response.status_code == 200

        # load_legal_search_results gets called twice in this view,
        # so this mocks the two different calls and then we assert they happend
        # http://stackoverflow.com/questions/7242433/asserting-successive-calls-to-a-mock-method
        calls = [
            mock.call(query='', query_type='advisory_opinions', 
                      ao_min_issue_date=ao_min_date, ao_category=['F', 'W'])
        ]
        load_legal_search_results.assert_has_calls(calls, any_order=True)
